front/src/modules/simulationResult/components/SpeedSpaceChart/helpers.ts

```diff
@@ -14,10 +14,78 @@ import type {
   SimulationResponseSuccess,
 } from 'applications/operationalStudies/types';
 import type { ReportTrainV2 } from 'common/api/osrdEditoastApi';
+import { TAG_COLORS } from 'modules/simulationResult/consts';
+import type { SpeedLimitTagValue } from 'modules/simulationResult/types';
 import { mmToKm, msToKmh, mToKm } from 'utils/physics';
 
 import { electricalProfilesDesignValues } from './consts';
 
+const getTag = (source?: SpeedLimitTagValue['source']): { name: string; color: string } => {
+  let name: string;
+  let color: string;
+
+  switch (source?.speed_limit_source_type) {
+    case 'given_train_tag':
+      name = source.tag;
+      color = TAG_COLORS.GIVEN_TRAIN;
+      break;
+    case 'fallback_tag':
+      name = source.tag;
+      color = TAG_COLORS.FALLBACK;
+      break;
+    case 'unknown_tag':
+      name = 'incompatible';
+      color = TAG_COLORS.INCOMPATIBLE;
+      break;
+    default:
+      name = 'missing_from_train';
+      color = TAG_COLORS.MISSING;
+  }
+
+  return { name, color };
+};
+
+export const formatSpeedLimitTags = (
+  rawSpeedLimitTags: SimulationResponseSuccess['mrsp'],
+  pathLength: number
+): LayerData<SpeedLimitTagValues>[] =>
+  rawSpeedLimitTags.values
+    .reduce((mergedPositions, value, index) => {
+      const start = index === 0 ? 0 : rawSpeedLimitTags.boundaries[index - 1];
+      const end =
+        index === rawSpeedLimitTags.values.length - 1
+          ? pathLength
+          : rawSpeedLimitTags.boundaries[index];
+
+      const { name, color } = getTag(value.source);
+
+      const newPosition = {
+        position: {
+          start: mmToKm(start),
+          end: mmToKm(end),
+        },
+        value: {
+          tag: name,
+          color,
+        },
+      };
+
+      const lastPosition = mergedPositions[mergedPositions.length - 1];
```

Comment:
I find it cleaner like that (and it forces us to check the undefined case, as you already did)

```suggestion
      const lastPosition = mergedPositions.at(-1);
```